### PR TITLE
Add cattle label resource + API; affinity constraints via env variables

### PR DESCRIPTION
**Pull request by sonchang**

### CODE REVIEW

File: code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AbstractAllocator.java [hunk 1]
```diff
@@ -326,9 +342,28 @@ protected void close(Iterator<AllocationCandidate> iter) {
...
         for (AllocationConstraintsProvider provider : allocationConstraintProviders) {
-            provider.appendConstraints(attempt, log, constraints);
+            if (requestedHostId == null || provider.isCritical()) {
+                provider.appendConstraints(attempt, log, constraints);
+            }
         }
+        Collections.sort(constraints, new Comparator<Constraint>() {
+            @Override
+            public int compare(Constraint o1, Constraint o2) {
+                if (o1 == o2) return 0;
```
Comment (ibuildthecloud) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AbstractAllocator.java]: Can you just explain what your doing here?  Are you sorting to get soft constraints last??

Comment (sonchang) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AbstractAllocator.java]: Yup.  Just sorting to get soft constraints last.  Also, I'm thinking the existing spread algorithm based on the compute free should prevent us from experiencing a bin-packing affect after soft constraints are removed.

Comment (ibuildthecloud) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AbstractAllocator.java]: Awesome, that's makes sense.

**Comment by sonchang** [comment]
@ibuildthecloud @alena1108 
Most of the changes are in the auto-generated jooq code.

File: code/framework/object/src/main/java/io/cattle/platform/object/meta/impl/DefaultObjectMetaDataManager.java [hunk 2]
```diff
@@ -234,19 +234,30 @@ protected void findMappings() {
...
-                String mapRightToLeftName = right.getRight().getPropertyName();
+                // override to obtain a more meaningful link name
                 register(getLinkNameOverride(entry.getKey().getSimpleName(), mapRightToLeftName, mapRightToLeftName), entry.getKey(), left, right);
-
-                String mapLeftToRightName = left.getRight().getPropertyName();
                 register(getLinkNameOverride(entry.getKey().getSimpleName(), mapLeftToRightName, mapLeftToRightName), entry.getKey(), right, left);
             } else {
-                register(schemaFactory.getSchema(right.getLeft()).getPluralName(), entry.getKey(), left, right);
-                register(schemaFactory.getSchema(left.getLeft()).getPluralName(), entry.getKey(), right, left);
+                register(getLinkNameOverride(
```
Comment (ibuildthecloud) [inline code/framework/object/src/main/java/io/cattle/platform/object/meta/impl/DefaultObjectMetaDataManager.java]: Can you explain what this is doing?  Making the name of something configurable?

**Comment by ibuildthecloud** [comment]
LGTM

@son should I merge or are you still working on any of this?

**Comment by sonchang** [comment]
@ibuildthecloud Go ahead and merge.  Thanks!

**Pull request merged**